### ISSUE TITLE
Clarification of RabbitMQ Queues utilized by servicecontrol error & audit instances

### ISSUE THREAD

**Issue by TraGicCode** [Particular/ServiceControl#1951]
**ServiceControl Version:**
4.4.1

**Audit Instance and Error Instance Names**
Particular.ServiceControl.Prod
Particular.ServiceControl.Prod.Audit

**Transport:**
RabbitMQ
Error Fowrarding is Enabled
Audit Forwarding is Enabled

**Rabbit Queues Created:**

Particular.ServiceControl.Prod
Particular.ServiceControl.Prod.Audit
Particular.ServiceControl.Prod.Audit.Errors
Particular.ServiceControl.Prod.AWS-PROD01
Particular.ServiceControl.Prod.Errors
Particular.ServiceControl.Prod.staging
audit.forwarding.prod
error.forwarding.prod

1. What is this queue used for `Particular.ServiceControl.Prod.Audit.Errors`?
1. What is this queue used for `Particular.ServiceControl.Prod.Errors`?
1. What is this queue used for `Particular.ServiceControl.Prod.staging`?
1. What is this queue used for `Particular.ServiceControl.Prod.AWS-PROD01`?

**Comment by danielmarbach** [comment]
Hi Michael

- `Particular.ServiceControl.Prod.Audit.Errors` and `Particular.ServiceControl.Prod.Errors` are the error queues of the instance command bus. For example certain types of ServicePulse commands do internally trigger a message that then gets handled asynchronously.
- `Particular.ServiceControl.Prod.staging` this is the queue used to stage messages that are retried and sent to a destination. This queue is only used if you retry a batch of messages. Messages will first be moved there and then from there to the final destination
- `Particular.ServiceControl.Prod.AWS-PROD01` not sure why this would be there honestly. Is there an active consumer on that queue on rabbitmq? Doesn't ring a bell to me so it could be something you are using?

**Comment by TraGicCode** [comment]
> Particular.ServiceControl.Prod.AWS-PROD01 not sure why this would be there honestly. Is there an active consumer on that queue on rabbitmq? Doesn't ring a bell to me so it could be something you are using?

I kind of already knew this one but its a leftover artifact from the old nservicebus days when the callback queue for a given machine was already created.  I'm going to clean this up manually on our side.

> Particular.ServiceControl.Prod.staging this is the queue used to stage messages that are retried and sent to a destination. This queue is only used if you retry a batch of messages. Messages will first be moved there and then from there to the final destination

Any idea why it just doesn't go to the service control created input queue ( Particular.ServiceControl.Prod ) ? Maybe to prevent getting clogged up behind other messages?

> Particular.ServiceControl.Prod.Audit.Errors and Particular.ServiceControl.Prod.Errors are the error queues of the instance command bus. For example certain types of ServicePulse commands do internally trigger a message that then gets handled asynchronously.

I believe messages that fail to get forwarded to forwarding queues if they are setup also land here.  Any idea how these messages can get retried if they end up here?  Is that feature even possible?

**Comment by danielmarbach** [comment]
Here is further details about the design of the staging approach should you wish to dig deeper

https://github.com/Particular/ServiceControl/blob/master/docs/bulk-retries-design.md

The approach is written so that we avoid duplicates even for transports that don't support SendsWithAtomicReceive. Doing that over a dedicated queue for that purpose was the best we could come up with at that time.

> I believe messages that fail to get forwarded to forwarding queues if they are setup also land here. Any idea how these messages can get retried if they end up here? Is that feature even possible?

We do pass a copy of those messages that fail ingestion to the error queue. You should not attempt retry messages from there because SC will overwrite the error info with it's own exceptions and stuff
i.e. The FailedQ header will get replaced with the error/audit queue that the message was being ingested from so retrying will just send the message back to there

So in short the error queue of SC has not a lot of purpose. The only thing that could really end up there might be a custom check report that you might want to retry manually. So more an additional safety net.

**Comment by TraGicCode** [comment]
Hey @danielmarbach ,

Thanks for the details response.  Appreciate the work and help you are doing! 👍
